### PR TITLE
User can view allocations

### PR DESCRIPTION
**Pull request by rotimi-babalola**
## What does this PR do?
User can view all allocations on the `/allocations` route

## How should this be manually tested?
Open the app and navigate to the `/allocations` route from the side menu

## What are the relevant pivotal tracker stories?
#158492837

## Any background context you want to add?
N/A

## Important notes
N/A

## Packages installed
N/A

## Related PRs branch | PR (branch_name) | (pr_link)
Branch | PR
------ | ------
other_pr_production | [link]()
other_pr_master | [link]()
## Todos
- [x] Raise PR
- [ ] Test
- [ ] Documentation
## Screenshots
screen shot 2018-06-22 at 19 00 14

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -71,4 +71,4 @@
     "stylelint": "^9.3.0",
     "stylelint-config-standard": "^18.2.0"
   }
-}
+}
```
Comment (sachungo) [inline package.json]: Add newline, please. It will be easier to set your editor to automatically add it for you

File: src/__test__/actions/allocations.actions.test.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  });
+
+  it('should dispatch LOADING_ALLOCATIONS and LOAD_ALLOCATIONS_FAILURE when allocations are not loaded', () => {
+    mock.onGet('allocations/').reply(401);
+    return store.dispatch(loadAllocationsAction()).then(() => {
+      expect(store.getActions()[0].type).toEqual(LOADING_ALLOCATIONS);
+      expect(store.getActions()[1].type).toEqual(LOAD_ALLOCATIONS_FAILURE);
+    });
+  });
+});
```
Comment (sachungo) [inline src/__test__/actions/allocations.actions.test.js]: Add newline, please

File: src/_components/SideMenuComponent.jsx [hunk 3]
```diff
@@ -84,7 +87,11 @@ class SideMenuComponent extends Component {
 
 SideMenuComponent.propTypes = {
   children: PropTypes.node.isRequired,
-  title: PropTypes.string.isRequired
+  title: PropTypes.string
 };
 
+SideMenuComponent.defaultProps = {
+  title: '',
```
Comment (sachungo) [inline src/_components/SideMenuComponent.jsx]: No need for the comma dangle, please delete

File: src/_mock/allocations.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+      created_at: "2018-05-07T16:49:58.211873Z",
+    },
+    {
+      asset: "DSGHDSGHSDGS",
+      current_owner: 9,
+      previous_owner: 9,
+      created_at: "2018-05-14T15:07:33.182482Z",
+    },
+  ]
+}
```
Comment (sachungo) [inline src/_mock/allocations.js]: Newline please

File: src/components/AllocationsComponent.jsx [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+  }
+
+  handlePaginationChange = (event, { activePage }) => {
+    this.setState({
+      activePage,
+      offset: (activePage - 1) * this.state.limit,
+    }, () => this.setTableContent());
+  }
+
+  isEmptyAllocations = () => this.props.allAllocations.length === 0;
```
Comment (sachungo) [inline src/components/AllocationsComponent.jsx]: This method is redundant. [Lodash](https://github.com/AndelaOSP/art-dashboard/blob/develop/package.json#L15) exists in the repo so please use the [isEmpty](https://lodash.com/docs/4.17.10#isEmpty) method

File: src/components/AllocationsComponent.jsx [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+          </p>
+        </Container>
+      );
+    }
+    return (
+      <Container>
+        <Header content="All Allocations" />
+        <Table celled>
+          <Table.Header>
+            {/* @TODO make the header cells more flexible depending on response from API */}
```
Comment (sachungo) [inline src/components/AllocationsComponent.jsx]: When will this comment be actioned?

File: src/components/AllocationsComponent.jsx [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  handlePaginationChange = (event, { activePage }) => {
+    this.setState({
+      activePage,
+      offset: (activePage - 1) * this.state.limit,
+    }, () => this.setTableContent());
+  }
+
+  isEmptyAllocations = () => this.props.allAllocations.length === 0;
+
+  showAllocations = () => {
```
Comment (sachungo) [inline src/components/AllocationsComponent.jsx]: I usually discourage separating parts of the render function into other functions in a class. This pattern ends up making the file to be extra long as people will tend to add more functions in the same class in future to render new additions to the markup.

If you must do this for the component to render, then it's time you move the logic to it's own component. Otherwise, map through the state in the render function

File: src/components/AllocationsComponent.jsx [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        </Table>
+      </Container>
+    );
+  }
+
+  render() {
+    return (
+      <SideMenuComponent>
+        {
+          this.display()
```
Comment (sachungo) [inline src/components/AllocationsComponent.jsx]: Why separate the markup to `display` function? Read my previous comment

File: src/components/AllocationsComponent.jsx (hunk 6 above)
Comment (rotimi-babalola) [reply src/components/AllocationsComponent.jsx]: I think I forgot about that 😃 I'll delete the comment but it's something worth thinking about 🤔

File: src/components/AllocationsComponent.jsx [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+                <Table.HeaderCell>Previous Owner</Table.HeaderCell>
+                <Table.HeaderCell>Created At</Table.HeaderCell>
+              </Table.Row>
+            </Table.Header>
+
+            <Table.Body>
+              {
+                this.state.currentAllocations.map((allocation, index) => (
+                  <TableRowComponent
+                    key={index}
```
[the remark this answers is not in the document]
Comment (rotimi-babalola) [reply src/components/AllocationsComponent.jsx]: To fix this, I think the backend team needs to add IDs to the allocations. Currently there are no IDs and that is why I am using the index

File: src/components/AllocationsComponent.jsx [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+      (this.state.activePage * this.state.limit));
+    // format date
+    const dateOptions = {
+      weekday: 'long',
+      year: 'numeric',
+      month: 'long',
+      day: 'numeric',
+    }
+    currentAllocations.map(el => {
+      return el.created_at = new Date(el.created_at).toLocaleDateString('en-US', dateOptions);
```
Comment (theterminalguy) [inline src/components/AllocationsComponent.jsx]: There is no point for the assignment

File: src/components/AllocationsComponent.jsx [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    });
+    this.setState({
+      currentAllocations: currentAllocations,
+    });
+  }
+
+  handlePaginationChange = (event, { activePage }) => {
+    this.setState({
+      activePage,
+      offset: (activePage - 1) * this.state.limit,
```
Comment (theterminalguy) [inline src/components/AllocationsComponent.jsx]: Put the expression for calculating offset into its own method.

File: src/components/AllocationsComponent.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (theterminalguy) [reply src/components/AllocationsComponent.jsx]: For now, you can use the`created_at` field as the key for the array. You should be able to get pass the houndci errors

File: src/components/AllocationsComponent.jsx [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+                ))
+              }
+            </Table.Body>
+
+            <Table.Footer>
+              <Table.Row>
+                <Table.HeaderCell colSpan="4">
+                  {
+                    (_.isEmpty(this.props.allAllocations)) ? '' :
+                      <Pagination
```
Comment (theterminalguy) [inline src/components/AllocationsComponent.jsx]: Proper indentation, please.

**Comment by theterminalguy** [comment]
Please fix all hound ci complaint and all failing test. Also, go through the notes I left on the PR @rotimi-babalola

File: src/components/AllocationsComponent.jsx (hunk 10 above)
Comment (rotimi-babalola) [reply src/components/AllocationsComponent.jsx]: It doesn't format the date properly if I don't do the assignment. Or is there something I am missing? @theterminalguy

File: src/components/AllocationsComponent.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (rotimi-babalola) [reply src/components/AllocationsComponent.jsx]: The problem with this is I've already formatted the `created_at` field to string format at the time of render. So this field is no longer unique anymore for each allocation. cc @theterminalguy

**Comment by rotimi-babalola** [comment]
Will fix the merge conflicts

File: .eslintrc [hunk 13]
```diff
@@ -24,7 +24,10 @@
       "error",
       "always"
     ],
-    "comma-dangle": ["error", "never"],
+    "comma-dangle": [
```
Comment (sachungo) [inline .eslintrc]: out of curiosity - why this change? The array is too short for multiline values

File: .eslintrc [hunk 14]
```diff
@@ -77,10 +80,15 @@
     ],
     "no-restricted-syntax": "off",
     "no-prototype-builtins": "off",
+    "no-return-assign": "off",
+    "no-param-reassign": "off",
     "no-underscore-dangle": "off",
     "no-use-before-define": "off",
     "object-curly-newline": "off",
-    "react/jsx-indent": [2, 2],
+    "react/jsx-indent": [
```
Comment (sachungo) [inline .eslintrc]: Same here

File: src/__test__/AllocationComponent.test.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+import expect from 'expect';
+
+import { AllocationsComponent } from '../components/AllocationsComponent';
+import allocations from '../_mock/allocations';
+
+describe('Renders <Allocations/> component', () => {
+  const props = {
+    loadAllocationsAction: jest.fn(),
+    allAllocations: [],
+    isLoading: true,
```
[the remark this answers is not in the document]
Comment (sachungo) [reply src/__test__/AllocationComponent.test.js]: Please delete the trailing comma

File: src/__test__/actions/allocations.actions.test.js [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+import thunk from 'redux-thunk';
+
+import constants from '../../_constants';
+import allocations from '../../_mock/allocations';
+import loadAllocationsAction from '../../_actions/allocations.actions';
+
+const {
+  LOAD_ALLOCATIONS_SUCCESS,
+  LOAD_ALLOCATIONS_FAILURE,
+  LOADING_ALLOCATIONS,
```
[the remark this answers is not in the document]
Comment (sachungo) [reply src/__test__/actions/allocations.actions.test.js]: Delete trailing comma

File: src/__test__/actions/allocations.actions.test.js [hunk 17]
```diff
@@ -0,0 +1,59 @@
+import expect from 'expect';
+import MockAdapter from 'axios-mock-adapter';
+import axios from 'axios';
+import configureMockStore from 'redux-mock-store';
+import thunk from 'redux-thunk';
+
+import constants from '../../_constants';
```
[the remark this answers is not in the document]
Comment (sachungo) [reply src/__test__/actions/allocations.actions.test.js]: @rotimi-babalola Delete the blank line above this to fix `Expected 1 empty line after import statement not followed by another import` houndci comment

File: src/__test__/reducers/allocations.reducer.test.js [hunk 18]
```diff
@@ -0,0 +1,43 @@
+import expect from 'expect';
+import allocationReducer from '../../_reducers/allocations.reducer';
+import allocations from '../../_mock/allocations';
+import constants from '../../_constants';
+
+const {
+  LOAD_ALLOCATIONS_SUCCESS,
+  LOAD_ALLOCATIONS_FAILURE,
+  LOADING_ALLOCATIONS,
```
[the remark this answers is not in the document]
Comment (sachungo) [reply src/__test__/reducers/allocations.reducer.test.js]: Delete trailing comma

File: src/__test__/reducers/allocations.reducer.test.js [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+} = constants;
+
+const initialState = {
+  allAllocations: [],
+  isLoading: false,
```
[the remark this answers is not in the document]
Comment (sachungo) [reply src/__test__/reducers/allocations.reducer.test.js]: Same here

File: src/components/AllocationsComponent.jsx [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+  componentDidUpdate(prevProps) {
+    if (this.props.allAllocations.length !== prevProps.allAllocations.length) {
+      this.setTableContent();
+    }
+  }
+
+  getTotalPages = () => Math.ceil(this.props.allAllocations.length / this.state.limit);
+
+  setTableContent = () => {
+    const currentAllocations = this.props.allAllocations.slice(this.state.offset,
```
Comment (sachungo) [inline src/components/AllocationsComponent.jsx]: Write this as:

```
const currentAllocations = this.props.allAllocations.slice(
  this.state.offset,
  (this.state.activePage * this.state.limit)
);
```

for readability

File: src/components/AllocationsComponent.jsx [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+  }
+
+  render() {
+    if (this.props.isLoading) {
+      return (
+        <SideMenuComponent>
+          <LoaderComponent size="large" dimmerStyle={{ height: '100vh' }} />
+        </SideMenuComponent>
+      );
+    } else if (!this.props.isLoading && _.isEmpty(this.props.allAllocations)) {
```
Comment (sachungo) [inline src/components/AllocationsComponent.jsx]: Use of `else if`  is unnecessary due to the return statements. Having it as
```
if (!this.props.isLoading && _.isEmpty(this.props.allAllocations)) {
  ...
}
```
will suffice

File: .eslintrc (hunk 13 above)
Comment (rotimi-babalola) [reply .eslintrc]: I think my editor probably formatted it like this after saving this file @sachungo

File: .eslintrc (hunk 14 above)
Comment (rotimi-babalola) [reply .eslintrc]: Probably my editor's fault again @sachungo 😃

File: src/components/LoginComponent.jsx [hunk 22]
```diff
@@ -26,6 +26,7 @@ class LoginComponent extends React.Component {
         localStorage.setItem('art-prod-web-token', idToken);
         this.props.history.push('/dashboard');
       });
+      ToastMessage.success({ message: 'Welcome to ART' });
```
Comment (sachungo) [inline src/components/LoginComponent.jsx]: This message is unnecessary. The user will be directed to the dashboard so they'll know that they are already logged in
